### PR TITLE
Color picker updates

### LINKED ISSUE #7878: ColorPicker resets to red when dragged to edge
### Environment Information 
 
- __Package version(s)__: 6.126.2
- __Browser and OS versions__: (fill this out if relevant)

### Please provide a reproduction of the bug in a codepen:
 

See demo page:

https://developer.microsoft.com/en-us/fabric#/components/colorpicker
 
#### Actual behavior:

When using the color picker and dragging around in the color square, the hue slider slightly changes/there is a wobble. Then if you hit the left or bottom edge, the whole window changes to red instead of whatever hue you were trying to use. The hue slider however stays the same.

#### Expected behavior:
 
There should be no wobble on the hue slider when moving and hitting an edge should not reset the color window to red.

#### Priorities and help requested:
 
Are you willing to submit a PR to fix?No
 
Requested priority: Normal

### PR DESCRIPTION
**Pull request by ecraig12345**
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #7878
- [x] Include a change request file using `$ npm run change`

#### Description of changes

The ColorPicker previously only allowed `props.color` to be a string, which isn't ideal because the conversion from color object to string and back is imprecise. So this PR adds the option of passing an `IColor` object as `props.color`. 

I also added a standard change handler prop `onChange: (ev: React.SyntheticEvent<HTMLElement>, color: IColor) => void` to the ColorPicker, which is called only when the user triggered the change (not when a props change occurred). 

Not directly related to the issue, I combined the text field rendering and change logic for the hex text field with the RGB text fields. This makes the code smaller but more importantly helps get ready for another change I'd like to make that adds better handling of invalid values.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/7957)

### CODE REVIEW

File: packages/office-ui-fabric-react/src/components/ColorPicker/ColorPicker.base.tsx [hunk 1]
```diff
@@ -142,60 +127,79 @@ export class ColorPickerBase extends BaseComponent<IColorPickerProps, IColorPick
...
+        // props update (which is not very useful)
+        if (props.onColorChanged) {
+          props.onColorChanged(newColor.str, newColor);
+        }
+      });
     }
   }
 }
+
+function _getColorFromProps(props: IColorPickerProps) {
```
Comment (natalieethell) [inline packages/office-ui-fabric-react/src/components/ColorPicker/ColorPicker.base.tsx]: return type?

File: packages/office-ui-fabric-react/src/components/ColorPicker/ColorRectangle/ColorRectangle.base.tsx [hunk 2]
```diff
@@ -103,19 +107,19 @@ export class ColorRectangleBase extends BaseComponent<IColorRectangleProps, ICol
...
     ev.preventDefault();
     ev.stopPropagation();
   };
 
   private _onMouseUp = (ev: React.MouseEvent<HTMLElement>): void => {
     this._events.off();
-
-    this.setState({
-      isAdjusting: false,
-      origin: undefined
```
Comment (natalieethell) [inline packages/office-ui-fabric-react/src/components/ColorPicker/ColorRectangle/ColorRectangle.base.tsx]: just curious, what did these do before?

Comment (ecraig12345) [reply packages/office-ui-fabric-react/src/components/ColorPicker/ColorRectangle/ColorRectangle.base.tsx]: They were set but never accessed anywhere. Probably artifacts of some previous implementation.

File: apps/fabric-website-resources/src/components/pages/ColorsPage.tsx [hunk 3]
```diff
@@ -444,14 +438,14 @@ export class ColorsPage extends BaseComponent<{}, IColorsPageState> {
   private _baseColorSlotPicker = (baseSlot: BaseSlots, title: string): JSX.Element => {
     let colorChangeTimeout: number;
 
-    function _onColorChanged(newColor: string): void {
+    const onChange = (ev: any, newColor: IColor): void => {
```
Comment (KevinTCoughlin) [inline apps/fabric-website-resources/src/components/pages/ColorsPage.tsx]: Are we unable to type these `ev` parameters? I thought they might be `React.MouseEvent<HTMLElement>` glancing at the code.

**Comment by cliffkoh** [comment]
@ecraig12345 before you merge this, could you help me test out this following edge case with AutoMerge?

1. Add label AutoMerge
2. Close PR
3. Re-open PR

Previously, AutoMerge failed in this case but I reasonably confident it is fixed.. just need to repro this scenario though to validate the fix. 

Thanks so much in advance!

cc @Vitalius1 @kkjeer

File: .github/CODEOWNERS [hunk 4]
```diff
@@ -73,7 +73,7 @@ Callout/  @joschect
 # Check/
 # Checkbox/
 ChoiceGroup/   @srideshpande
-# ColorPicker/
+ColorPicker/ @ecraig12345
```
Comment (dzearing) [inline .github/CODEOWNERS]: thank you @ecraig12345 !!! awesome taking ownership

File: apps/fabric-website-resources/src/components/pages/ColorsPage.tsx (hunk 3 above)
Comment (ecraig12345) [reply apps/fabric-website-resources/src/components/pages/ColorsPage.tsx]: That was more laziness than anything else...I can add a type.

**Comment by ecraig12345** [comment]
@cliffkoh Tried to do your steps but just realized I did it in the wrong order (close, open, label). Going to try again now as close, label, open but wanted to let you know in case having done it wrong once messes up the experiment...

**Comment by cliffkoh** [comment]
Order only matters so the bot doesn't immediately merge your PR when you first add the label :P

**Comment by cliffkoh** [comment]
BTW it needs to be in a mergable state first (i.e. has approvals)

File: packages/office-ui-fabric-react/src/components/ColorPicker/ColorPicker.base.tsx [hunk 5]
```diff
@@ -32,27 +25,32 @@ export class ColorPickerBase extends BaseComponent<IColorPickerProps, IColorPick
...
   };
-  private _rgbaLabels: { [K in keyof IRGB]?: string };
+  private _textLabels: { [K in keyof IRGBHex]?: string };
 
   constructor(props: IColorPickerProps) {
     super(props);
 
+    this._warnDeprecations({
+      onColorChanged: 'onChange'
+    });
```
Comment (cliffkoh) [inline packages/office-ui-fabric-react/src/components/ColorPicker/ColorPicker.base.tsx]: :+1:

File: packages/office-ui-fabric-react/src/components/ColorPicker/ColorPicker.deprecated.test.tsx [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    if (wrapper) {
+      wrapper.unmount();
+      wrapper = undefined;
+    }
+  });
+
+  it('respects color prop change', () => {
+    const onChange = jest.fn();
+    wrapper = mount(<ColorPicker color="#abcdef" onColorChanged={onChange} componentRef={colorPickerRef} />);
+
```
Comment (cliffkoh) [inline packages/office-ui-fabric-react/src/components/ColorPicker/ColorPicker.deprecated.test.tsx]: :+1:

**Comment by ecraig12345** [comment]
Closing in another bug repro attempt :)

**Comment by natalieethell** [comment]
Hm, the alpha slider doesn't seem to be working properly. Note that the alpha value changes, but the other color values at the bottom aren't changing.
ezgif com-video-to-gif 2

**Comment by cliffkoh** [comment]
Thanks @natalieethell for testing! You should block the PR so it isn't merged!

**Comment by ecraig12345** [comment]
Good catch @natalieethell -- I'll look into that.

**Comment by ecraig12345** [comment]
This thing soooooooo badly needs tests........ :(

File: apps/fabric-website-resources/src/components/pages/ColorsPage.tsx (hunk 3 above)
Comment (KevinTCoughlin) [reply apps/fabric-website-resources/src/components/pages/ColorsPage.tsx]: No worries, just figured might as well. Lmk if you want me to take a pass at it. Great work btw!

File: packages/office-ui-fabric-react/src/components/ColorPicker/ColorPicker.deprecated.test.tsx [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    wrapper.setProps({ color: '#AEAEAE' });
+    expect(colorPickerRef.current!.color.hex).toEqual('aeaeae');
+    // weird but preserves existing behavior
```
Comment (kenotron) [inline packages/office-ui-fabric-react/src/components/ColorPicker/ColorPicker.deprecated.test.tsx]: Can we be more specific about what is "weird" about this behavior?

Comment (ecraig12345) [reply packages/office-ui-fabric-react/src/components/ColorPicker/ColorPicker.deprecated.test.tsx]: Calling onChange when props are updated is weird. I'll add a comment.

**Comment by ecraig12345** [comment]
@natalieethell Fixed the bug and added a test :)
